### ISSUE TITLE
Unable to authenticate Desktop App through gitlab OAUTH2 Login since 9.1.x

### ISSUE THREAD

**Issue by wech71** [mattermost/desktop#2906]
#### Summary
Since upgrade to Server 9.1.0 (and 9.1.1) it does no longer work to log in trough gitlab authentication (oauth2) to the desktop app.

After downgrade to 9.0.0 it works again.

#### Steps to reproduce

1. Self-hosted Mattermost Server behind nginx reverse proxy on /subpath
2. Open Windows Desktop app (5.4.0, 5.5.0 or 5.5.1 - makes no difference)
3. Ensure to be logged out of Mattermost
4.  Klick "gitlab" on mattermost login page and notice difference in behaviour:
*   -> In 9.1.x The gitlab login page opens in an external browser window, 
*   -> In 9.0.0 The gitlab login page opens inside the desktop app
5.  After logging in to gitl gitlab, offers the possibility to continue in the browser (wich works) or in the desktop app trough the mattermost: pseudo-protocol (mattermost://subdomain.domain.tld/chat/login/desktop?client_token=xxxx&server_token=yyyy)
6. Klick Desktop App button

#### Expected behavior
Mattermost desktop app should be logged in and display the chat-channel

#### Observed behavior (that appears unintentional) 
The "Channels" tab is being shown with white background and a gray hyperlink to subdomain.domain.tld/subpath, after reload with Ctrl+R the mattermost login page with username/password and the "gitlab" button (same as in Step 4) is being displayed.

#### Possible fixes

Workaround: Downgraded Server to 9.0.0

I could not find any dubious requests in the Mattermost Developer Tools or nginx reverse-proxy except some 401 errors like the request for
   https://subdomain.domain.tld/subpath/api/v4/cloud/subscription/self-serve-status

**Comment by devinbinnie** [comment]
Transferred back to desktop repo as it turns out there was an issue with the app misinterpreting deep links for subpaths. PR is up to fix :)

**Comment by marjoleintamis** [comment]
I'm not sure if this is the same issue but I'm tagging onto it because I think it might be. I ran into a similar thing when trying to log in to a server via GitLab. Some specifics: 

- The server version is 9.2.2
- Running Mattermost Desktop version 5.5.1 commit db8465b
- Logging in refers me to the GitLab setup in Firefox where I'm already logged in via 2FA. The site refers me back to the Mattermost app, which gives me a blank screen with only the Channels tab and the server name on it. Clicking on the server name sends me back to the login page.
- I also tried this in Chrome and Edge, where I had to log in fresh. Same result.
- Ctrl-R also reloads to the login page.

Logs: 

```
Preload initialized
index.js:150 Loading plugin jitsi, version 2.0.1
index.js:150 Loading plugin mattermost-standup, version 0.1.0
index.js:150 Loading plugin com.github.matterpoll.matterpoll, version 1.6.1
index.js:150 Loading plugin com.mattermost.calls, version 0.20.0
index.js:150 Loading plugin focalboard, version 7.11.4
index.js:136 Loaded plugin mattermost-standup, version 0.1.0
index.js:136 Loaded plugin jitsi, version 2.0.1
index.js:136 Loaded plugin com.github.matterpoll.matterpoll, version 1.6.1
index.js:136 Loaded plugin com.mattermost.calls, version 0.20.0
plugins/jitsi/api/v1/config:1     Failed to load resource: the server responded with a status of 401 ()
focalboard_bc5a129749169408_bundle.js:2 [1700033888.89] OctoClient baseURL: https://site.domain.nl/chat/plugins/focalboard
index.js:136 Loaded plugin focalboard, version 7.11.4
com.mattermost.calls_ed1c001e0ba11dd4_bundle.js:2 com.mattermost.calls: loading translations file for locale 'en'
api/v4/teams?page=0&per_page=200&include_total_count=false&exclude_policy_constrained=false:1     Failed to load resource: the server responded with a status of 401 ()
Metropolis-SemiBold.woff2:1     Failed to load resource: the server responded with a status of 404 ()
api/v4/teams?page=0&per_page=200&include_total_count=false&exclude_policy_constrained=false:1     Failed to load resource: the server responded with a status of 401 ()
api/v4/teams?page=0&per_page=200&include_total_count=false&exclude_policy_constrained=false:1     Failed to load resource: the server responded with a status of 401 ()
DevTools failed to load source map: Could not load content for https://site.domain.nl/chat/plugins/jitsi/external_api.min.map: HTTP error: status code 404, net::ERR_HTTP_RESPONSE_CODE_FAILURE
```

If this is a different issue, I can make a separate issue of it. Thanks!

**Comment by derzinn** [comment]
While we wait for an update, the workaround is to manually register the MIME type:

```
xdg-mime default com.mattermost.Desktop.desktop x-scheme-handler/mattermost
```

Tested on Fedora Workstation 39 with Mattermost Desktop Version 5.5.1 (Flatpak).

**Comment by wech71** [comment]
Just a short feedback: After upgrading to Mattermost Desktop 5.6.0.RC1 and Mattermost Server 9.2.3 everything seems to work fine now :-)

**Comment by grisuthedragon** [comment]
I updated my installation from 8.1.x to 9.3.0 on the server side and using the 5.6.0 client on Linux ( Ubuntu 22.04 ) and we still ran into the same issue. 
 
* I am using mattermost on a subpath `/mattermost` via nginx reverse proxy.
* Self hosted 
* oauth2 with Gitllab. 
* Desktop app from the *deb package.
* Mime-Type registered as described by @derzinn 

The desktop app redirects to the browser, where I can authenticate against Gitlab, but then I does not switch back to the app.  So the fix in #2907 did not solve the problem properly.

**Comment by Leny1996** [comment]
@devinbinnie looks like this issue is still present, we're facing the same problem when using Gitlab Oauth2 setup

**Comment by ja11sop** [comment]
Please reopen this issue as clearly it still persists. We now have failures now across our entire estate and across all platforms including Windows, Kubuntu, Manjaro and Debian on both Firefox and Chrome. We're contemplating ditching Mattermost after many good years since we can no longer use the desktop application.

MM 5.7.0
Server 9.5.0 (self hosted)
Gitlab Oauth2 for authentication - redirect to desktop application now never works

**Comment by marvinl1986** [comment]
This issue still persists.

@devinbinnie we still have this issue:

Client: MM 5.7.0-832 (Ubuntu 22.04)
Server 9.7.2 (self hosted)
Gitlab Oauth2 for authentication - redirect to desktop application does not work

**Comment by ToppDev** [comment]
For me it worked with Firefox, but not with Chrome/Brave (they always open another browser window instead of Mattermost)

Client: Version 5.7.0 commit: e999e2c (NixOS unstable)
Server: 9.8.0 (self-hosted)
Gitlab Oauth2 authentication

Mime-Types: `~/.config/mimeapps.list`
```
[Added Associations]
text/html=firefox.desktop
x-scheme-handler/chrome=firefox.desktop
x-scheme-handler/http=firefox.desktop
x-scheme-handler/https=firefox.desktop
x-scheme-handler/mattermost=Mattermost.desktop

[Default Applications]
text/html=firefox.desktop
x-scheme-handler/chrome=firefox.desktop
x-scheme-handler/http=firefox.desktop
x-scheme-handler/https=firefox.desktop
x-scheme-handler/mattermost=Mattermost.desktop
```

`Mattermost.desktop `
```
[Desktop Entry]
Name=Mattermost
Comment=Mattermost Desktop application for Linux
Exec="/nix/store/dgrlq4nh0k6z03a4ff81w75fb7ws2b9c-mattermost-desktop-5.7.0/bin/mattermost-desktop" %U
Terminal=false
Type=Application
MimeType=x-scheme-handler/mattermost
Icon=/nix/store/dgrlq4nh0k6z03a4ff81w75fb7ws2b9c-mattermost-desktop-5.7.0/share/mattermost-desktop/app_icon.png
Categories=Network;InstantMessaging;
```

**Comment by perrelle** [comment]
The issue is still present for me.
system: Ubuntu 23.10

**Comment by devinbinnie** [comment]
Does this seem to be common for Firefox users? Is everyone still experiencing this issue using Firefox?
I know I've gotten it to work with Firefox before, but I think it requires some manual configuration.

If this is the case, I can revisit how I did it and see if I can create a quick walkthrough, and add to our docs.

**Comment by nightroxide** [comment]
I seem to have the same issue on Firefox. Furthermore, using chrome I can get all the way to the last step @shaneonabike gets to, but the mattermost app does not refresh, and I can only log in using the browser

**Comment by shaneonabike** [comment]
@nightroxide did you try refresh in the Mattermost app? Before I did that and it started working for chrome. Firefox no dice.

@devinbinnie yes even if it doesn't affect everyone providing some help would be amazing or better integration. Thanks for all you do for the community

**Comment by devinbinnie** [comment]
No problem :) will spend some time soon and get a guide together on how to address this.

FWIW I just set up a new Linux box today (for the first time in a while) and going through our APT repository as described [here](https://docs.mattermost.com/collaborate/install-desktop-app.html#install-and-update-the-mattermost-desktop-app) I had no problems logging in to the Desktop App with Firefox. This was using Kubuntu 24.04.

@shaneonabike Can you share your distribution and how you installed the Desktop App? (eg. Snap, via GitHub - and which package, APT, etc)

**Comment by shaneonabike** [comment]
No trouble.

+ Ubuntu 22.04
+ APT via deb https://deb.packages.mattermost.com stable main
+ Firefox 130.0.1

I really don't know if this makes a different but I use Containers. I'm not opening the token in any special container but the default (no container).

**Comment by devinbinnie** [comment]
It might, but I wouldn't think so. Whatever is happening seems to be stopping Firefox from being able to associate the `mattermost:` protocol to our app, since that should happen automatically with the Desktop App once it's loaded for the first time.

**Comment by devinbinnie** [comment]
So fun thing - this actually happened to me today using Kubuntu 24.04 when I went to modifiy the command line arguments to add the arguments to force native Wayland mode.

I noticed that the new `.desktop` file created in `~/.local/share/applications` was missing the `MimeType`, despite the fact that the original in `/usr/share/applications` does have it.

For anyone still running into this issue, make sure if you have both `.desktop` files that they both contain the line `MimeType=x-scheme-handler/mattermost;`

**Comment by devinbinnie** [comment]
Just tried with Firefox again, I do get the prompt below to open the app, but then nothing happens. So this does seem to be some strangeness with Firefox:
![image](https://github.com/user-attachments/assets/1e1e8b13-e535-4422-abbd-2c648ed5006e)

**Comment by devinbinnie** [comment]
@shaneonabike Do you see a line like this showing up in your Mattermost desktop logs?
```
[2024-09-27 11:17:06.033] [debug] [App.App] handleAppSecondInstance [
  '/opt/Mattermost/mattermost-desktop',
  '--enable-features=UseOzonePlatform',
  '--ozone-platform=wayland',
  '--allow-file-access-from-files',
  '--enable-sandbox',
  'mattermost://community.mattermost.com/login/desktop?client_token=[token]&server_token=[token]'
]
```

**Comment by cyberm8** [comment]
@devinbinnie Hello,

On 5.10 the issue is still there.

Desktop environment: Debian 12 with KDE Wayland.
We added the MIME type manually.

The observed behavior is as follows: the Mattermost application refreshes after successful authentication on Gitlab (the link works fine from the browser), very briefly, we see the Mattermost interface, then the application returns to “you are now logged in”.

Restarting the application allows you to use it afterwards, which is better than not being able to use it, but it assumes you've manually added the mime type.

![image](https://github.com/user-attachments/assets/f147240a-da49-4106-bdd8-248f5eac9618)

[mattermost_desktop_510_debian_12_wayland_kde_gitlab_oauth.log](https://github.com/user-attachments/files/17816566/mattermost_desktop_510_debian_12_wayland_kde_gitlab_oauth.log)

**Comment by devinbinnie** [comment]
@cyberm8 I'm not sure what the problem is, so is it able to log you in to the Desktop App successfully, but it doesn't forward you to the application after login until you refresh? Or is it that the MIME type isn't manually added?

I can see in the logs that the app is trying to push you to the application correctly:
`[2024-11-19 15:16:47.209] [debug] [ViewManager] handleBrowserHistoryPush 3 /common/channels/jw41cw5j5prr9dkyq99ncyrary__smupeqggxfn63p6eathytywxmy`

I'm also seeing some weird behaviour here:
```
[2024-11-19 15:16:47.047] [debug] [PermissionsManager] doPermissionRequest openExternal {
  externalURL: 'mattermost://mattermost.lan.HIDDEN.fr/login/desktop/?client_token=[REDACTED]&server_token=[REDACTED]',
  isMainFrame: true,
  requestingUrl: 'https://mattermost.lan.HIDDEN.fr/login/desktop/?client_token=[REDACTED]&server_token=[REDACTED]'
}
```

It looks like the app is trying to open the deep link within the Desktop App. What server version are you on?

**Comment by cyberm8** [comment]
Ah, actually there are two related problems.

First, to answer the question about the environment:
- Desktop 5.10
- Server 9.11.1 (we're updating it to the latest version this evening)
- Debian 12 KDE Wayland

The first problem is the one that justified the creation of this github issue. Specifically, after Gitlab authentication, Firefox is unable to open mattermost:// links. Reading the thread, the problem seemed to have been solved, but this is not the case a priori.
I can confirm that there is no problem with Chrome/Chromium.

So we added the mime type manually.

Authentication goes further, as Firefox now knows how to interpret mattermost:// links, and indeed everything seems to work fine (the interface with messages and channels is displayed) but half a second later the application displays the “you're already logged in” page again. The workaround is to restart the app.

I think a video might help.

To sum up, MIME type must be added manually (maybe we did not add it properly) and the app must be restarted in order to access the app.
